packages/core/src/typegenTypes.ts

```diff
@@ -0,0 +1,262 @@
+> = TResolvedTypesMeta extends TypegenEnabled
+  ? IsNever<
+      Values<
+        {
+          [K in keyof TMissingImplementations]: TMissingImplementations[K];
+        }
+      >
+    > extends true
+    ? true
+    : false
+  : true;
+
+export type MarkAllImplementationsAsProvided<
+  TResolvedTypesMeta
+> = TResolvedTypesMeta & {
+  missingImplementations: {
+    actions: never;
+    delays: never;
+    guards: never;
+    services: never;
+  };
+};
+
+// we don't even have to do that much here, technically, because `T & unknown` is equivalent to `T`
+// however, this doesn't display nicely in IDE tooltips, so let's fix this
+type MergeWithInternalEvents<TIndexedEvents, TInternalEvents> = TIndexedEvents &
+  // alternatively we could consider using key remapping in mapped types for this in the future
+  // in theory it would be a single iteration rather than two
+  Pick<TInternalEvents, Exclude<keyof TInternalEvents, keyof TIndexedEvents>>;
+
+export type ResolveTypegenMeta<
+  TTypesMeta extends TypegenConstraint,
+  TEvent extends EventObject,
+  TAction extends BaseActionObject
+> = TTypesMeta extends TypegenEnabled
+  ? TTypesMeta & {
+      indexedActions: IndexByType<TAction>;
+      indexedEvents: MergeWithInternalEvents<
+        IndexByType<TEvent>,
+        Prop<TTypesMeta, 'internalEvents'>
+      >;
+    }
+  : TypegenDisabled;
+
+export type TypegenMachineOptionsActions<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingActions = Prop<TResolvedTypesMeta, 'eventsCausingActions'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>,
+  TIndexedActions = Prop<TResolvedTypesMeta, 'indexedActions'>
+> = {
+  [K in keyof TEventsCausingActions]?:
+    | ActionObject<
+        TContext,
+        Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>
+      >
+    | ActionFunction<
+        TContext,
+        Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>,
+        Cast<Prop<TIndexedActions, K>, BaseActionObject>
+      >;
+};
+
+export type TypegenMachineOptionsDelays<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingDelays = Prop<TResolvedTypesMeta, 'eventsCausingDelays'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>
+> = {
+  [K in keyof TEventsCausingDelays]?: DelayConfig<
+    TContext,
+    Cast<Prop<TIndexedEvents, TEventsCausingDelays[K]>, EventObject>
+  >;
+};
+
+export type TypegenMachineOptionsGuards<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingGuards = Prop<TResolvedTypesMeta, 'eventsCausingGuards'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>
+> = {
+  [K in keyof TEventsCausingGuards]?: ConditionPredicate<
+    TContext,
+    Cast<Prop<TIndexedEvents, TEventsCausingGuards[K]>, EventObject>
+  >;
+};
+
+export type TypegenMachineOptionsServices<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingServices = Prop<TResolvedTypesMeta, 'eventsCausingServices'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>,
```

Comment:
When a service does not have any `onDone` attribute and it's declared as a promise, it returns `Promise<never>` - which is not correct - it should return `Promise<unknown>`

Reply:
Actually, I might be able to fix this on the typegen side - I don't think we're providing all the `internalEvents` that we should be.

Reply:
ah, this makes sense - if there is no transition defined for a particular `invoke` then you don't have that on the list of `.transitions`

Reply:
Fixed this on the typegen side, but still erroring in the same way, discussed on Discord